### PR TITLE
Switch to building soljson.js with base64 encoded embedded wasm binary.

### LINKED ISSUE #2090: Compile soljson to Webassembly
It would make sense to generate a target, with the same API as emscripten, for Webassembly.

Limitations:
- llvm/clang must be rolled by ourselves, because the target is experimental
- browsers don't have it enabled by default

Guide on compiling clang for wasm: https://gist.github.com/axic/c369b413ae782362f52c6b42bc1e3d7f

### PR DESCRIPTION
**Pull request by ekpyron**
Closes #2090
Depends on #7832

Should build soljson.js and soljson.wasm for synchronous loading.

### CODE REVIEW

**Comment by ekpyron** [comment]
External tests of course fail, because we'd need to copy ``soljson.wasm`` around for them.

**Review by axic — CHANGES_REQUESTED** [review]
We should actually build both asm.js and wasm.

**Comment by ekpyron** [comment]
I'd say before we go ahead with this at all we should test the wasm version in remix.

**Comment by axic** [comment]
@ekpyron I think we should actually postpone this after 0.6.0 and leave enough time to experiment/figure this one out

We made great progress and know this is possible at least.

**Comment by ekpyron** [comment]
> @ekpyron I think we should actually postpone this after 0.6.0 and leave enough time to experiment/figure this one out
> 
> We made great progress and know this is possible at least.

Yep, I agree - this is mainly opened as PR to have CI build it and not with the intention to merge it anytime soon :-).

**Comment by ekpyron** [comment]
@axic You want to see something funny :-)? Then look at the last commit - if we in fact base64 encode the wasm file and prepend it to the auxiliary ``soljson.js`` with some decoding logic, we end up with a 7MB drop in replacement for regular ``soljson.js``, which can be loaded into remix-alpha.

**Comment by ekpyron** [comment]
I just played around with it a bit... it seems to work perfectly and seems magnitudes faster on (artificially) larger contracts...
The integrated loading logic is not very good, though, apparently... it tries to deduce an URL for ``soljson.wasm``, but doesn't manage (if I add logging to it it still tries to load plain ``soljson.wasm``, which raises an "Invalid URL" error).
Manually providing ``Module["locateFile"]`` would probably work as well, though, but I haven't properly tested it so far.

**Comment by ekpyron** [comment]
Yeah, apparently when loaded from a remix worker soljson.js will end up getting a "scriptDirectory" of ``blob:http://remix-alpha.ethereum.org/<SOME-UUID>`` only and can't deduce the location of soljson.wasm based on that... but manually providing ``Module["locateFile"]`` for it should indeed work.

**Comment by ekpyron** [comment]
Also: it required quite a bit of tinkering, but I managed to build a drop-in replacement for 0.4.25's soljson.js (containing base64 encoded wasm) and checked in remix that it indeed loads and works as expected and doesn't exhibit the stack height problem.
It'll be quite some work to do this cleanly and in a well-documented way, but I think this will make for a nice solution for the browser problems.

**Comment by axic** [comment]
> Then look at the last commit - if we in fact base64 encode the wasm file and prepend it to the auxiliary soljson.js with some decoding logic, we end up with a 7MB drop in replacement for regular soljson.js, which can be loaded into remix-alpha.

Cool idea! I was worried the base64 encoding would increase it a lot and/or slow down loading too much (due to decoding).

How about [ascii85 / encoding85](https://en.wikipedia.org/wiki/Ascii85)? That uses more symbols (e.g. more compact), but has still a "JSON compatible" character set, e.g. no escaping needed in Javascript (the RFC-1924 version, specifically).

**Comment by axic** [comment]
Also @ekpyron I think this feature of merging wasm into it could be an upstream patch, at somem point.

**Comment by ekpyron** [comment]
I thought about it and we can try, but I don't think ascii85 is worth it - in fact we might do the opposite and also try hex - I think simple and fast conversion might be more important than data density - the whole thing is likely to be zipped for https resp. http/2 transfer anyways (so we should also check what's most compact when compressed, not only uncompressed).

**Comment by ekpyron** [comment]
> Also @ekpyron I think this feature of merging wasm into it could be an upstream patch, at somem point.

Given the stack height problems in the asm.js version and the vastly increased compilatiion speed, plus still smaller file size than asm.js, plus pretty ubiquitous wasm support in browsers these days, maybe even rather soon...

**Comment by axic** [comment]
> I thought about it and we can try, but I don't think ascii85 is worth it - in fact we might do the opposite and also try hex - I think simple and fast conversion might be more important than data density - the whole thing is likely to be zipped for https resp. http/2 transfer anyways (so we should also check what's most compact when compressed, not only uncompressed).

Good point. I was concerned about binary sizes given we put so many into `solc-bin` (which btw was over a 1 GB a year ago, probably over 2 GB now), but perhaps I shouldn't be and be more worried on loading speed.

**Comment by ekpyron** [comment]
I'm surprised that github hasn't complained about that repo yet :-)... We should in fact consider building them in github actions and publishing them in github packages in the solidity repo instead - that would also remove the need for travis (or anyone) having push permissions anywhere... but that's a whole different thing, independent of the wasm build itself - I don't think where and how it's going to be hosted should be the prime concern when building it.

**Comment by axic** [comment]
So actually it seems the asm.js is well compressible so the entire repo in git packs is only 400 Mb.

However unpacked it is 7.6 Gb.

**Comment by ekpyron** [comment]
At least for zip-compressing an individual release I get the following:
```
-rw-r--r-- 1 daniel users 2927510 30. Nov 12:44 soljson.asm.js.nightly.develop.zip
-rw-r--r-- 1 daniel users 2133889 30. Nov 12:43 soljson.combined.base64.zip
-rw-r--r-- 1 daniel users 1786462 30. Nov 12:53 soljson.combined.hex.zip
-rw-r--r-- 1 daniel users 1502348 30. Nov 12:43 soljson.separate.js.and.wasm.zip
```

File: scripts/travis-emscripten/base64.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
+/*\
```
Comment (chriseth) [inline scripts/travis-emscripten/base64.js]: License?

File: test/solcjsTests.sh [hunk 2]
```diff
@@ -51,7 +52,9 @@ DIR=$(mktemp -d)
     # Replace soljson with current build
     echo "Replacing soljson.js"
     rm -f soljson.js
-    cp "$SOLJSON" soljson.js
+	rm -f soljson.wasm
```
Comment (chriseth) [inline test/solcjsTests.sh]: indentation

File: scripts/travis-emscripten/base64.js (hunk 1 above)
Comment (ekpyron) [reply scripts/travis-emscripten/base64.js]: Yeah, good question... Judging from https://developer.mozilla.org/en-US/docs/MDN/About#Copyrights_and_licenses about "Code samples and snippets", *probably* ``public domain (CC0) No licensing notice is necessary`` applies...

**Comment by ekpyron** [comment]
I just rebased and switched this to only building one soljson.js with base64-encoded wasm embedded in it - let's see if it still passes all tests.

If that works, the following remains to be done:
- changelog entry and documentation of the change and the new requirement for running soljson.js
- verify the travis build, etc., works (I think that's not all done yet)
- consider using uglifyjs to minify the base64 decoding code and to automatically remove the redundant loading logic from soljson.js (as @axic suggested)
- make sure the license for the base64 decoding code is fine, or replace, or switch to hex after all

**Comment by ekpyron** [comment]
I just checked: ``uglifyjs`` won't help and the way the loading code is embedded in the js part makes it a bit tricky to remove it... I'm checking if there is an option for emscripten to not even generate it, but it doesn't look like it - so in the end I think it's easiest to just ignore those few lines of dead code.

**Comment by ekpyron** [comment]
Haha, ok, the PR just shrunk to a fraction of its size and should still work - in fact emscripten had an option ``SINGLE_FILE=1`` that does the whole base64-encoded embedding for us. I remembered something like that last time, but hadn't found it for some reason - but yeah, no need to do any of it manually on our end.

File: cmake/EthCompilerSettings.cmake [hunk 3]
```diff
@@ -111,8 +111,10 @@ if (("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU") OR ("${CMAKE_CXX_COMPILER_ID}" MA
 			# Export the Emscripten-generated auxiliary methods which are needed by solc-js.
 			# Which methods of libsolc itself are exported is specified in libsolc/CMakeLists.txt.
 			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s EXTRA_EXPORTED_RUNTIME_METHODS=['cwrap','addFunction','removeFunction','UTF8ToString','lengthBytesUTF8','_malloc','stringToUTF8','setValue']")
```
Comment (axic) [inline cmake/EthCompilerSettings.cmake]: No need to export `_malloc` anymore.

File: cmake/EthCompilerSettings.cmake [hunk 4]
```diff
@@ -111,8 +111,10 @@ if (("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU") OR ("${CMAKE_CXX_COMPILER_ID}" MA
...
-			# Do not build as a WebAssembly target - we need an asm.js output.
-			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s WASM=0")
+			# Build standalone webassembly for synchronous loading.
+			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s WASM=1 -s WASM_ASYNC_COMPILATION=0")
+			# Output a single js file with embedded wasm.
+			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s SINGLE_FILE=1")
```
Comment (axic) [inline cmake/EthCompilerSettings.cmake:119]: I would suggest to either separate all three options or put all three on a single line.

File: scripts/travis-emscripten/build_emscripten.sh [hunk 5]
```diff
@@ -54,14 +54,6 @@ fi
 
 WORKSPACE=/root/project
 
-# Increase nodejs stack size
```
Comment (axic) [inline scripts/travis-emscripten/build_emscripten.sh:57]: This was needed for compilation and **not** for the resulting binary. Is it safe to remove it now?

Comment (ekpyron) [reply scripts/travis-emscripten/build_emscripten.sh:57]: Yep, not needed anymore.

Comment (axic) [reply scripts/travis-emscripten/build_emscripten.sh:57]: Sorry, updated the comment, it was **not** needed for the binary, but the compilation process :)

Comment (ekpyron) [reply scripts/travis-emscripten/build_emscripten.sh:57]: Yep, I know :-) - but it's not needed anymore in any case - the wasm build goes through without it.

**Review by chriseth — APPROVED** [review]

**Pull request merged**